crates/extensions/c8y_auth_proxy/src/server.rs

```diff
@@ -0,0 +1,307 @@
+        .post(respond_to)
+        .put(respond_to)
+        .patch(respond_to)
+        .delete(respond_to)
+        .options(respond_to);
+    let app = Router::new()
+        .route("/c8y", handle.clone())
+        .route("/c8y/*path", handle)
+        .with_state(state);
+    Ok(axum::Server::try_bind(&(address, port).into())
+        .into_diagnostic()
+        .wrap_err_with(|| format!("binding to port {port}"))?
+        .serve(app.into_make_service()))
+}
+
+#[derive(Clone)]
+pub(crate) struct AppState {
+    pub target_host: Arc<str>,
+    pub token_manager: SharedTokenManager,
+}
+
+impl FromRef<AppState> for TargetHost {
+    fn from_ref(input: &AppState) -> Self {
+        Self(input.target_host.clone())
+    }
+}
+
+impl FromRef<AppState> for SharedTokenManager {
+    fn from_ref(input: &AppState) -> Self {
+        input.token_manager.clone()
+    }
+}
+
+#[derive(Clone)]
+struct TargetHost(Arc<str>);
+
+impl fmt::Display for TargetHost {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+async fn respond_to(
+    State(TargetHost(host)): State<TargetHost>,
+    retrieve_token: State<SharedTokenManager>,
+    path: Option<Path<String>>,
+    method: Method,
+    headers: HeaderMap<HeaderValue>,
+    small_body: crate::body::PossiblySmallBody,
+) -> Response {
+    let path = match &path {
+        Some(Path(p)) => p.as_str(),
+        None => "",
+    };
+
+    if path.ends_with(".js") || path.starts_with("apps/") {
+        return Response::builder()
+            .status(StatusCode::FORBIDDEN)
+            .body(<_>::default())
+            .unwrap();
+    }
+    let destination = format!("{host}/{path}",);
+
+    let mut token = retrieve_token.not_matching(None).await;
+
+    let client = reqwest::Client::new();
+    let (body, body_clone) = small_body.try_clone();
+    let send_request = |body, token| {
+        client
+            .request(method.to_owned(), destination.clone())
+            .headers(headers.clone())
+            .bearer_auth(&token)
+            .body(body)
+            .send()
+    };
+    let mut res = send_request(body, token.clone()).await.unwrap();
+
+    if res.status() == StatusCode::UNAUTHORIZED {
+        token = retrieve_token.not_matching(Some(&token)).await;
+        if let Some(body) = body_clone {
+            res = send_request(Body::from(body), token.clone()).await.unwrap();
+        }
+    }
+    let te_header = res.headers_mut().remove("transfer-encoding");
+    let headers = std::mem::take(res.headers_mut());
+    let mut build = Response::builder().status(res.status());
+    *build.headers_mut().unwrap() = headers;
+
+    let body = if te_header.map_or(false, |h| {
+        h.to_str().unwrap_or_default().contains("chunked")
+    }) {
+        axum::body::boxed(StreamBody::new(res.bytes_stream()))
+    } else {
+        axum::body::boxed(Full::new(res.bytes().await.unwrap()))
+    };
+
+    build.body(body).unwrap()
```

Comment:
If the body is sufficiently large that we don't cache it in memory via `PossiblySmallBody`, then the body is simply preserved as a stream and just passed straight to `reqwest`. I haven't done any specific tests of this, but there's no reason why this shouldn't work. The issue currently would be that it would still succumb to expired JWTs. We deal with that by retrying a request transparently, but if we haven't stored the (potententially very large) body in memory, we cannot retry the request (but if the client were to retry the request to the proxy that would then work).